torch/_inductor/codegen/wrapper.py

```diff
@@ -1189,9 +1189,16 @@ def define_user_defined_triton_kernel(self, kernel, configs, kwargs):
         compile_wrapper = IndentedBuffer()
         compile_wrapper.writeline(f"async_compile.triton({original_name!r}, '''")
 
-        from .triton import gen_common_triton_imports, TritonKernel
+        from .triton import (
+            define_global_constexpr_vars,
+            gen_common_triton_imports,
+            TritonKernel,
+        )
 
         compile_wrapper.splice(gen_common_triton_imports())
+        # constexpr vars captured by the kernel must be defined w/ annotations
+        constexpr_global_symbol_names, buffer = define_global_constexpr_vars(kernel)
```

Comment:
I was looking at this same issue and ended up resolving it independently by adding another branch to the check in the `traverse` method in `WrapperCodeGen`: https://github.com/alexbaden/pytorch/commit/cee200c2a5e593cd43a4635926f6d41b15587703#diff-18714c66f0b1f90de1bfd784bae5fab68a4e54d817556fefa5cdee2621cb06d0R1253

It's a little bit less code and seems to work with various permutations of the `test_triton_kernels` test, but I am new to Inductor and was curious if there were other pros and cons vs this implementation.

Reply:
This is a bit cleaner, but there are some problems with it.

1) We probably shouldn't codegen `symbol_name = tl.constexpr(value)` for any global (int, str, bool) since the source kernel in that case may fail to compile with triton in eager mode if there is not a `tl.constexpr` annotation. 

2) Due to the above we have to look at the annotations to determine if an object that is not an instance of `tl.constexpr` is okay to codegen as one. 

3) We should probably not codegen annotated variables as instances though. Since this can change the meaning of the users source code for the kernel. 

The original code for the test has the conditional `CONSTANT_NAME.value == STRING_CONSTANT_C`. This works if `STRING_CONSTANT_C` is defined with a `tl.constexpr` annotation but not if it is a `tl.constexpr` object (`STRING_CONSTANT_C = tl.constexpr("CONSTANT_C")`).

However, the annotations get a bit messed up as they pass through our interface so I need to work out a fix for that.

Reply:
I think the extra checks to make sure we aren't adding annotations that didn't exist before makes sense. However, it also looks like `traverse` function is the more natural place to add this functionality.